[ci] Python tests failing with error about numpy image

## Description

For the last week or so, I have seen a few Python CI jobs fail with an error like the following

```text
__________ ERROR collecting tests/python_package_test/test_sklearn.py __________
ImportError while importing test module '/Users/runner/work/LightGBM/LightGBM/tests/python_package_test/test_sklearn.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
../../../../miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/__init__.py:22: in <module>
    from . import multiarray
../../../../miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/multiarray.py:12: in <module>
    from . import overrides
../../../../miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/overrides.py:7: in <module>
    from numpy.core._multiarray_umath import (
E   ImportError: dlopen(/Users/runner/miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so, 2): Library not loaded: @rpath/libopenblas.dylib
E     Referenced from: /Users/runner/miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so
E     Reason: image not found

During handling of the above exception, another exception occurred:
../../../../miniconda/envs/test-env/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
../tests/python_package_test/test_sklearn.py:7: in <module>
    import numpy as np
../../../../miniconda/envs/test-env/lib/python3.9/site-packages/numpy/__init__.py:140: in <module>
    from . import core
../../../../miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/__init__.py:48: in <module>
    raise ImportError(msg)
E   ImportError: 
E   
E   IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
E   
E   Importing the numpy C-extensions failed. This error can happen for
E   many reasons, often due to issues with your setup or how NumPy was
E   installed.
E   
E   We have compiled some common reasons and troubleshooting tips at:
E   
E       https://numpy.org/devdocs/user/troubleshooting-importerror.html
E   
E   Please note and check the following:
E   
E     * The Python version is: Python3.9 from "/Users/runner/miniconda/envs/test-env/bin/python"
E     * The NumPy version is: "1.19.2"
E   
E   and make sure that they are the versions you expect.
E   Please carefully study the documentation linked above for further help.
E   
E   Original error was: dlopen(/Users/runner/miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so, 2): Library not loaded: @rpath/libopenblas.dylib
E     Referenced from: /Users/runner/miniconda/envs/test-env/lib/python3.9/site-packages/numpy/core/_multiarray_umath.cpython-39-darwin.so
E     Reason: image not found
```

I see the same stacktrace (or very similar) for all test modules, not just `test_sklearn.py`

```text
=========================== short test summary info ============================
ERROR ../tests/python_package_test/test_basic.py
ERROR ../tests/python_package_test/test_consistency.py
ERROR ../tests/python_package_test/test_dask.py
ERROR ../tests/python_package_test/test_dual.py
ERROR ../tests/python_package_test/test_engine.py
ERROR ../tests/python_package_test/test_plotting.py
ERROR ../tests/python_package_test/test_sklearn.py
ERROR ../tests/python_package_test/test_utilities.py
!!!!!!!!!!!!!!!!!!! Interrupted: 8 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 8 errors in 2.18s ===============================
```


I most recently saw this on the `sdist (macOS-latest, Python 3.9)` on GitHub Actions (https://github.com/microsoft/LightGBM/pull/4203/checks?check_run_id=2382402602).

## Reproducible example

This has been happening sporadically on CI jobs. In every case where I've encountered it, I have found that re-running the job fixes it.

## Environment info

LightGBM continuous integration.

## Additional Comments

I have seen this error on several different PRs, most of which did not touch the Python package, so I do not think it is related to any one open PR. For example, the build linked above is from #4203, which only has R package changes.

I see one similar issue on `numpy`'s issue tracker: https://github.com/numpy/numpy/issues/18663. In that issue, the author claimed that this error was solved by preferring `numpy` from `conda-forge` instead of Anaconda default channels.

I do see that the most recent version of `numpy` (1.20.2) was uploaded to conda-forge 21 days ago (https://anaconda.org/conda-forge/numpy)

![image](https://user-images.githubusercontent.com/7608904/115279700-a54e8c80-a10c-11eb-839c-1a9987d89159.png)

That version has not made it to the defaults channel yet. The last update of `numpy` there was version 1.19.2 more than 4 months ago (https://anaconda.org/anaconda/numpy).

![image](https://user-images.githubusercontent.com/7608904/115279775-bf886a80-a10c-11eb-892c-7152eafdf933.png)

I didn't mark this `blocking` since it seems to only happen occasionally.

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -104,8 +104,8 @@ else  # Linux
     fi
     ARCH=$(uname -m)
     if [[ $ARCH != "x86_64" ]] || [[ $TASK == "cuda" ]]; then
-        apt-get update
-        apt-get install --no-install-recommends -y \
+        apt update
+        apt install --no-install-recommends -y \
             graphviz
     else
         sudo apt-get update

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -81,6 +81,7 @@ else  # Linux
         mv $AMDAPPSDK_PATH/lib/x86_64/sdk/* $AMDAPPSDK_PATH/lib/x86_64/
         echo libamdocl64.so > $OPENCL_VENDOR_PATH/amdocl64.icd
     fi
+    ARCH=$(uname -m)
     if [[ $TASK == "cuda" ]]; then
         echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
         apt-get update
@@ -103,12 +104,17 @@ else  # Linux
         apt-get install --no-install-recommends -y \
             cmake
     else
-        sudo apt-get update
-        sudo apt-get install --no-install-recommends -y \
-            graphviz
+        if [[ $ARCH == "x86_64" ]]; then
+            sudo apt-get update
+            sudo apt-get install --no-install-recommends -y \
+                graphviz
+        else
+            apt-get update
+            apt-get install --no-install-recommends -y \
+                graphviz
     fi
     if [[ $SETUP_CONDA != "false" ]]; then
-        ARCH=$(uname -m)
+        
         if [[ $ARCH == "x86_64" ]]; then
             curl -sL -o conda.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
         else

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -25,13 +25,10 @@ else  # Linux
         echo 'debconf debconf/frontend select Noninteractive' | sudo debconf-set-selections
 
         sudo apt-get update
-        sudo apt-get install -y --no-install-recommends \
+        sudo apt-get install --no-install-recommends -y \
             software-properties-common
 
-        sudo add-apt-repository -y ppa:git-core/ppa
-        sudo apt-get update
-
-        sudo apt-get install -y --no-install-recommends \
+        sudo apt-get install --no-install-recommends -y \
             apt-utils \
             build-essential \
             ca-certificates \
@@ -64,12 +61,16 @@ else  # Linux
     fi
     if [[ $TASK == "mpi" ]]; then
         sudo apt-get update
-        sudo apt-get install --no-install-recommends -y libopenmpi-dev openmpi-bin
+        sudo apt-get install --no-install-recommends -y \
+            libopenmpi-dev \
+            openmpi-bin
     fi
     if [[ $TASK == "gpu" ]]; then
         sudo add-apt-repository ppa:mhier/libboost-latest -y
         sudo apt-get update
-        sudo apt-get install --no-install-recommends -y libboost1.74-dev ocl-icd-opencl-dev
+        sudo apt-get install --no-install-recommends -y \
+            libboost1.74-dev \
+            ocl-icd-opencl-dev
         cd $BUILD_DIRECTORY  # to avoid permission errors
         curl -sL -o AMD-APP-SDKInstaller.tar.bz2 https://github.com/microsoft/LightGBM/releases/download/v2.0.12/AMD-APP-SDKInstaller-v3.0.130.136-GA-linux64.tar.bz2
         tar -xjf AMD-APP-SDKInstaller.tar.bz2
@@ -84,6 +85,7 @@ else  # Linux
         apt-get update
         apt-get install --no-install-recommends -y \
             curl \
+            graphviz \
             libxau6 \
             libxext6 \
             libxrender1 \
@@ -99,6 +101,10 @@ else  # Linux
         apt-get update
         apt-get install --no-install-recommends -y \
             cmake
+    else
+        sudo apt-get update
+        sudo apt-get install --no-install-recommends -y \
+            graphviz
     fi
     if [[ $SETUP_CONDA != "false" ]]; then
         ARCH=$(uname -m)

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -106,7 +106,7 @@ else  # Linux
     else
         if [[ $ARCH != "x86_64" ]]; then
             yum update -y
-            yum install \
+            yum install -y \
                 graphviz
         else
             sudo apt-get update

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -81,13 +81,11 @@ else  # Linux
         mv $AMDAPPSDK_PATH/lib/x86_64/sdk/* $AMDAPPSDK_PATH/lib/x86_64/
         echo libamdocl64.so > $OPENCL_VENDOR_PATH/amdocl64.icd
     fi
-    ARCH=$(uname -m)
     if [[ $TASK == "cuda" ]]; then
         echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
         apt-get update
         apt-get install --no-install-recommends -y \
             curl \
-            graphviz \
             libxau6 \
             libxext6 \
             libxrender1 \
@@ -103,18 +101,18 @@ else  # Linux
         apt-get update
         apt-get install --no-install-recommends -y \
             cmake
+    fi
+    ARCH=$(uname -m)
+    if [[ $ARCH != "x86_64" ]] || [[ $TASK == "cuda" ]]; then
+        apt-get update
+        apt-get install --no-install-recommends -y \
+            graphviz
     else
-        if [[ $ARCH == "x86_64" ]]; then
-            sudo apt-get update
-            sudo apt-get install --no-install-recommends -y \
-                graphviz
-        else
-            apt-get update
-            apt-get install --no-install-recommends -y \
-                graphviz
+        sudo apt-get update
+        sudo apt-get install --no-install-recommends -y \
+            graphviz
     fi
     if [[ $SETUP_CONDA != "false" ]]; then
-        
         if [[ $ARCH == "x86_64" ]]; then
             curl -sL -o conda.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
         else

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -17,6 +17,7 @@ if [[ $OS_NAME == "macos" ]]; then
     if [[ $TASK == "swig" ]]; then
         brew install swig
     fi
+    brew install graphviz
     curl -sL -o conda.sh https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh
 else  # Linux
     if [[ $IN_UBUNTU_LATEST_CONTAINER == "true" ]]; then

--- a/.ci/setup.sh
+++ b/.ci/setup.sh
@@ -81,11 +81,13 @@ else  # Linux
         mv $AMDAPPSDK_PATH/lib/x86_64/sdk/* $AMDAPPSDK_PATH/lib/x86_64/
         echo libamdocl64.so > $OPENCL_VENDOR_PATH/amdocl64.icd
     fi
+    ARCH=$(uname -m)
     if [[ $TASK == "cuda" ]]; then
         echo 'debconf debconf/frontend select Noninteractive' | debconf-set-selections
         apt-get update
         apt-get install --no-install-recommends -y \
             curl \
+            graphviz \
             libxau6 \
             libxext6 \
             libxrender1 \
@@ -101,16 +103,16 @@ else  # Linux
         apt-get update
         apt-get install --no-install-recommends -y \
             cmake
-    fi
-    ARCH=$(uname -m)
-    if [[ $ARCH != "x86_64" ]] || [[ $TASK == "cuda" ]]; then
-        apt update
-        apt install --no-install-recommends -y \
-            graphviz
     else
-        sudo apt-get update
-        sudo apt-get install --no-install-recommends -y \
-            graphviz
+        if [[ $ARCH != "x86_64" ]]; then
+            yum update -y
+            yum install \
+                graphviz
+        else
+            sudo apt-get update
+            sudo apt-get install --no-install-recommends -y \
+                graphviz
+        fi
     fi
     if [[ $SETUP_CONDA != "false" ]]; then
         if [[ $ARCH == "x86_64" ]]; then

--- a/.ci/test.sh
+++ b/.ci/test.sh
@@ -104,14 +104,7 @@ if [[ $TASK == "swig" ]]; then
     exit 0
 fi
 
-conda install -q -y -n $CONDA_ENV cloudpickle dask distributed joblib matplotlib numpy pandas psutil pytest scikit-learn scipy
-
-# graphviz must come from conda-forge to avoid this on some linux distros:
-# https://github.com/conda-forge/graphviz-feedstock/issues/18
-conda install -q -y \
-    -n $CONDA_ENV \
-    -c conda-forge \
-        python-graphviz
+conda install -q -y -n $CONDA_ENV cloudpickle dask distributed joblib matplotlib numpy pandas psutil pytest python-graphviz scikit-learn scipy
 
 if [[ $OS_NAME == "macos" ]] && [[ $COMPILER == "clang" ]]; then
     # fix "OMP: Error #15: Initializing libiomp5.dylib, but found libomp.dylib already initialized." (OpenMP library conflict due to conda's MKL)

--- a/.ci/test.sh
+++ b/.ci/test.sh
@@ -104,7 +104,8 @@ if [[ $TASK == "swig" ]]; then
     exit 0
 fi
 
-conda install -q -y -n $CONDA_ENV cloudpickle dask distributed joblib matplotlib numpy pandas psutil pytest python-graphviz scikit-learn scipy
+conda install -q -y -n $CONDA_ENV cloudpickle dask distributed graphviz joblib matplotlib numpy pandas psutil pytest scikit-learn scipy
+pip install graphviz  # python-graphviz from Anaconda is not allowed to be installed with Python 3.9
 
 if [[ $OS_NAME == "macos" ]] && [[ $COMPILER == "clang" ]]; then
     # fix "OMP: Error #15: Initializing libiomp5.dylib, but found libomp.dylib already initialized." (OpenMP library conflict due to conda's MKL)

--- a/.ci/test.sh
+++ b/.ci/test.sh
@@ -104,7 +104,7 @@ if [[ $TASK == "swig" ]]; then
     exit 0
 fi
 
-conda install -q -y -n $CONDA_ENV cloudpickle dask distributed graphviz joblib matplotlib numpy pandas psutil pytest scikit-learn scipy
+conda install -q -y -n $CONDA_ENV cloudpickle dask distributed joblib matplotlib numpy pandas psutil pytest scikit-learn scipy
 pip install graphviz  # python-graphviz from Anaconda is not allowed to be installed with Python 3.9
 
 if [[ $OS_NAME == "macos" ]] && [[ $COMPILER == "clang" ]]; then

